src/server/server.zig

```diff
@@ -23,12 +25,64 @@ const command = @import("command/_command.zig");
 pub const WorldEditData = struct {
 	selectionPosition1: ?Vec3i = null,
 	selectionPosition2: ?Vec3i = null,
-	clipboard: ?main.blueprint.Blueprint = null,
+	clipboard: ?Blueprint = null,
+	undoHistory: History = .{},
+	redoHistory: History = .{},
+
+	const History = struct {
+		const DLList = std.DoublyLinkedList(Value);
+		changes: DLList = .{},
+		maxCapacity: u32 = 16,
+
+		const Value = struct {
+			blueprint: Blueprint,
+			position: Vec3i,
+			message: []const u8,
+
+			pub fn init(blueprint: Blueprint, position: Vec3i, message: []const u8) Value {
+				return .{.blueprint = blueprint, .position = position, .message = main.globalAllocator.dupe(u8, message)};
+			}
+			pub fn deinit(self: Value) void {
+				main.globalAllocator.free(self.message);
+				self.blueprint.deinit(main.globalAllocator);
+			}
+		};
+		pub fn deinit(self: *History) void {
```

Comment:
There is conceptual difference between `clear` and `deinit` hence I decided to have them separate despite both having currently same implementation.